src/Microsoft.AspNetCore.Authentication/AuthenticationHandler.cs

```diff
@@ -56,193 +56,104 @@ protected string CurrentUri
+        /// <param name="scheme"></param>
+        /// <param name="context"></param>
+        /// <returns></returns>
+        public virtual Task InitializeAsync(AuthenticationScheme scheme, HttpContext context)
         {
-            if (options == null)
+            if (scheme == null)
             {
-                throw new ArgumentNullException(nameof(options));
+                throw new ArgumentNullException(nameof(scheme));
             }
-
             if (context == null)
             {
                 throw new ArgumentNullException(nameof(context));
             }
 
-            if (logger == null)
-            {
-                throw new ArgumentNullException(nameof(logger));
-            }
-
-            if (encoder == null)
-            {
-                throw new ArgumentNullException(nameof(encoder));
-            }
-
-            Options = options;
+            Scheme = scheme;
             Context = context;
-            OriginalPathBase = Request.PathBase;
-            OriginalPath = Request.Path;
-            Logger = logger;
-            UrlEncoder = encoder;
-
-            RegisterAuthenticationHandler();
-
-            Response.OnStarting(OnStartingCallback, this);
 
-            if (ShouldHandleScheme(AuthenticationManager.AutomaticScheme, Options.AutomaticAuthenticate))
+            Options = OptionsSnapshot.Get(Scheme.Name) ?? new TOptions();
+            if (!Options.Initialized)
             {
-                InitializeResult = await HandleAuthenticateOnceAsync();
-                if (InitializeResult?.Skipped == true || InitializeResult?.Handled == true)
-                {
-                    return;
-                }
-
-                if (InitializeResult?.Failure != null)
+                lock (Options.InitializeLock)
                 {
-                    Logger.AuthenticationSchemeNotAuthenticatedWithFailure(Options.AuthenticationScheme, InitializeResult.Failure.Message);
-                }
-
-                var ticket = InitializeResult?.Ticket;
-                if (ticket?.Principal != null)
-                {
-                    Context.User = SecurityHelper.MergeUserPrincipal(Context.User, ticket.Principal);
-                    Logger.UserPrinicpalMerged(Options.AuthenticationScheme);
+                    if (!Options.Initialized)
+                    {
+                        InitializeOptions();
+                        Options.Initialized = true;
+                    }
                 }
             }
-        }
-
-        protected string BuildRedirectUri(string targetPath)
-        {
-            return Request.Scheme + "://" + Request.Host + OriginalPathBase + targetPath;
-        }
 
-        private static async Task OnStartingCallback(object state)
-        {
-            var handler = (AuthenticationHandler<TOptions>)state;
-            await handler.FinishResponseOnce();
-        }
+            // REVIEW: should we validate options only once inside of the lock?
```

Comment:
I think it needs to be outside. If the options are wrong then you need to fail every request, not just the first one.